librarian/data/meta/processors.py

```diff
@@ -0,0 +1,411 @@
+
+            {
+                '/path/original': {
+                    'path': '/path/1',
+                    'type': FILE OR FOLDER,
+                    'content_types': CONTENT TYPE BITMASK,
+                    'metadata': {
+                        'en': {...},
+                        'de': {...},
+                    }
+                },
+                '/path/redirected/if/any': {
+                    'path': '/path/2',
+                    'type': FILE OR FOLDER,
+                    'content_types': CONTENT TYPE BITMASK,
+                    'metadata': {
+                        'en': {...},
+                        'de': {...},
+                    }
+                }
+            }
+        """
+        # get path to which the found metadata should be connected
+        path = self.get_path()
+        # get existing metadata structure for either the original path or the
+        # path to which the metadata is being redirected
+        data_for_path = self.data.get(path, {})
+        # add generic data related to file system objects generally
+        self._add_fs_data(data_for_path)
+        # add extracted metadata
+        self._add_metadata(data_for_path)
+        # put back updated or newly created structure
+        self.data[path] = data_for_path
+
+    def deprocess(self):
+        """
+        Called when a meta entry is deleted from the database. Subclasses may
+        implement this method if special cleanup is needed.
+        """
+        pass
+
+    @classmethod
+    def is_entry_point(cls, new, old=None):
+        """
+        Processors handling types of more complex structure can elect a main
+        file which serves as a default entry point when opening a folder with
+        a specific facet. Returning ``True`` will tell the caller that the
+        passed in ``path`` can be used as the entry point. ``new`` should be
+        the path to the new candidate for election, and ``old`` should be the
+        existing one, which could affect the decision.
+        """
+        return False
+
+    @classmethod
+    def can_process(cls, path):
+        """
+        Return whether the processor can handle a given ``path``.
+        """
+        (_, ext) = os.path.splitext(path)
+        ext = ext[1:].lower()
+        return ext in cls.EXTENSIONS
+
+    @classmethod
+    def for_path(cls, path):
+        """
+        Return all the applicable processors for a given ``path``.
+        """
+        cls._subclasses = cls._subclasses or cls.subclasses()
+        return (proc_cls for proc_cls in cls._subclasses
+                if proc_cls.can_process(path))
+
+    @classmethod
+    def for_type(cls, content_type):
+        """
+        Return all the applicable processors for a given ``content_type``.
+        """
+        for proc_cls in cls.subclasses():
+            if proc_cls.name == content_type:
+                return proc_cls
+        raise RuntimeError("No processor found for the given content type.")
```

Comment:
1. The exception message should mention the content type
2. Might not make sense to raise an exception. It could just return `None`